### ISSUE TITLE
rope_scaling_type is a no-op on transformers 5: inv_freq and logits are byte-identical after "Long-context enabled"

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#793]
## The defect

`training.rope_scaling_type` (linear / dynamic / yarn / longrope / llama3) changes nothing about the model that trains. `trainer/sft.py` applies it **after** `from_pretrained`, by mutating `model.config`:

```

src/soup_cli/trainer/sft.py

1359    self.model = AutoModelForCausalLM.from_pretrained(cfg.base, **model_kwargs)

...

1367    # Long-context — apply RoPE scaling after model load

1368    if tcfg.rope_scaling_type:

1371        rope_config = apply_long_context_config(self.model.config, ...)

1377        if rope_config:

1378            console.print("[green]Long-context enabled:[/] RoPE ... scaling to ... tokens")

```

On transformers 5 the rotary embedding computes `inv_freq` and `attention_scaling` once, in `__init__`, from the config it was constructed with. Writing `config.rope_scaling` afterwards is never read again. The run prints "Long-context enabled" and trains with the base model's original RoPE.

Three further problems in the same path:

1. **The YaRN tunables are dropped.** `apply_long_context_config` (`utils/long_context.py:365-409`) calls `get_rope_scaling_config(scaling_type=..., target_length=..., original_length=...)` and never passes `yarn_factor` / `yarn_attn_factor` / `yarn_beta_fast` / `yarn_beta_slow`, although `get_rope_scaling_config` accepts all four (`:286-289`). The schema validates them (`config/schema.py:2775-2800`) and `_validate_yarn_fields_require_yarn_type` (`:3559-3562`) says it exists to avoid "silently dropping the values". `docs/peft-and-efficiency.md:127-139` shows a worked example setting all four.

2. **The dict it writes corrupts a saved config.** It uses the legacy `"type"` key and replaces `rope_parameters` wholesale, losing `rope_theta`. After a save/reload round trip, a `rope_theta=500000` model comes back with `rope_theta=10000` (reproduction below). Any path that serialises the model config, such as full fine-tuning with `lora.r: 0`, which ends in `trainer.save_model` (`sft.py:1972`), would write that.

3. **Only SFT reads the field.** `grep -rn rope_scaling_type src/soup_cli` finds `trainer/sft.py`, the `commands/train.py:1229` warning pre-flight, and the schema. `task: pretrain`, the natural task for context extension, accepts the field and ignores it entirely.

Also, `docs/peft-and-efficiency.md:142` says that omitting `rope_scaling_type` on a Llama 3.1 base makes `apply_long_context_config` auto-pick `llama3`. `sft.py:1368` only calls it when the field is set, so the auto-detect branch (`long_context.py:392-399`) is unreachable from `soup train`.

## Reproduction

Tiny random-init Llama, no download (transformers 5.17.0):

```python

m = LlamaForCausalLM(LlamaConfig(vocab_size=128, hidden_size=64, intermediate_size=128,

        num_hidden_layers=1, num_attention_heads=4, num_key_value_heads=4,

        max_position_embeddings=512)).eval()

inv0 = m.model.rotary_emb.inv_freq.clone(); out0 = m(ids).logits

apply_long_context_config(m.config, target_length=2048, rope_scaling_type=kind, model_name="x")

```

```

linear rope_config: {'type': 'linear', 'factor': 4.0}

  inv_freq changed: False attention_scaling: 1.0 -> 1.0 rope_type attr: default

  logits changed: False

yarn rope_config: {'type': 'yarn', 'factor': 4.0, 'original_max_position_embeddings': 512}

  inv_freq changed: False attention_scaling: 1.0 -> 1.0 rope_type attr: default

  logits changed: False

dynamic rope_config: {'type': 'dynamic', 'factor': 4.0}

  inv_freq changed: False attention_scaling: 1.0 -> 1.0 rope_type attr: default

  logits changed: False

control (linear x4 at construction) inv_freq differs from default: True

```

Save/reload of the mutated config (base built with `rope_theta=500000.0`):

```

fresh rope_parameters: {'rope_theta': 500000.0, 'rope_type': 'default'}

after apply rope_parameters: {'type': 'linear', 'factor': 4.0}

saved config.json rope keys: {'rope_parameters': {'factor': 4.0, 'rope_type': 'linear', 'type': 'linear'}}

reloaded rope_parameters: {'factor': 4.0, 'rope_type': 'linear', 'type': 'linear', 'rope_theta': 10000.0}

```

Moving the current dict before construction as-is would not work either:

```

construct with {'type': 'linear', 'factor': 4.0} failed: KeyError 'rope_type'

construct with {'rope_type': 'linear', 'factor': 4.0, 'rope_theta': 500000.0} -> inv_freq changed: True rope_type: linear

```

## Why it matters

Users who want long context are exactly the ones who set this, and the shipped configs do it for them: the `llama3.1-8b-longctx` recipe (`recipes/catalog.py:937`, `rope_scaling_type: yarn`) and `src/soup_cli/templates/longcontext.yaml:28` (`dynamic`, `max_length: 131072`). Those runs train at the long `max_length` on positions the model's RoPE was never extended for. Nothing fails, and the console confirms a feature that did not happen.

## Proposed fix

- Build the rope parameters **before** loading. Load `AutoConfig`, set `rope_parameters` in transformers-5 form (`rope_type`, and keep the existing `rope_theta` and any other keys), set `max_position_embeddings`, then `from_pretrained(cfg.base, config=config, ...)`.

- Forward `yarn_factor` / `yarn_attn_factor` / `yarn_beta_fast` / `yarn_beta_slow` from `TrainingConfig` into `get_rope_scaling_config`.

- Apply it in `pretrain.py` too, or reject `rope_scaling_type` at parse for tasks that do not read it.

- Either call the auto-detect path when the field is unset on a Llama 3.1 base, or correct the docs sentence.

## Acceptance criteria

- [ ] An effect test, not a dict test: build a tiny Llama through the real load path with `rope_scaling_type: linear` and assert `rotary_emb.inv_freq` differs from the unscaled model. It must fail on current code.

- [ ] A YaRN test asserting that `yarn_beta_fast` / `yarn_attn_factor` change `inv_freq` / `attention_scaling`.

- [ ] A save/reload test asserting `rope_theta` survives.

- [ ] `task: pretrain` either honours the field or refuses it with a message that names it.

- [ ] `docs/peft-and-efficiency.md` matches the behaviour.

---

**Related, filed the same day:**

- #806 — use_flash_attn and use_liger are read only by SFT's text path, yet validated on every task and switched on ...

**Comment by Amix29** [comment]
I’m taking this. I’ll reproduce the no-op through the real SFT load path first, then move RoPE configuration before model construction, preserve existing rope parameters such as rope_theta, forward the YaRN tunables, and make pretrain behavior explicit. Tests will assert an effect on the instantiated rotary embedding plus save/reload preservation; verification stays CPU/macOS and does not depend on CUDA.

**Comment by MakazhanAlpamys** [comment]
Fixed on `main` by @Amix29 in #931, merged as `2dd84c24`.

RoPE scaling was applied after the model was constructed, so the instantiated rotary module and the reloaded `config.json` both carried the default rather than the configured scaling. It is now configured before construction, and the ordering is atomic: the nested-`rope_parameters` refusal raises before `max_position_embeddings` is written and before the config reaches `from_pretrained`.

Four mutations, all discriminating on behaviour rather than source text — the assertions read the **instantiated** rotary module and the **reloaded** `config.json`, not the kwargs. Deleting the four `yarn_*` kwargs re-arms the consumer ratchet by name, which is what proves the allowlist retirement is a retirement and not a weakening.

494 passed on branch and merged tree; 2796 passed, 49 skipped across every file touching the SFT and pretrain wrappers.

Credit to @Amix29.
